grails-app/services/org/pih/warehouse/importer/InventoryImportDataService.groovy

```diff
@@ -0,0 +1,199 @@
+                    command.errors.reject(
+                            "error.lotAndExpiryControl.required",
+                            "Row ${rowIndex}: Both lot number and expiry date are required for the '${product.productCode} ${product.name}' product."
+                    )
+                }
+
+                def expirationDate = null
+                try {
+                    if (row.expirationDate) {
+                        if (row.expirationDate instanceof String) {
+                            expirationDate = dateFormatter.parse(row.expirationDate)
+                            calendar.setTime(expirationDate)
+                            expirationDate = calendar.getTime()
+                        } else if (row.expirationDate instanceof Date) {
+                            expirationDate = row.expirationDate
+                        } else if (row.expirationDate instanceof LocalDate) {
+                            expirationDate = row.expirationDate.toDate()
+                        } else {
+                            expirationDate = row.expirationDate
+                            command.warnings[index] << "Expiration date '${row.expirationDate}' has unknown format ${row?.expirationDate?.class}"
+                        }
+                        // Minimum date is either configured or we use the epoch date
+                        Date minExpirationDate = grailsApplication.config.getProperty("openboxes.expirationDate.minValue", Date.class, new Date(0L))
+                        if (minExpirationDate > expirationDate) {
+                            command.errors.reject("Expiration date for item ${row.productCode} is not valid. Please enter a date after ${minExpirationDate.getYear()+1900}.")
+                        }
+
+                        row.isNewExpirationDate = inventoryItem?.expirationDate && expirationDate != inventoryItem.expirationDate
+
+                        if (expirationDate <= new Date()) {
+                            command.warnings[index] << "Expiration date '${row.expirationDate}' is not valid"
+                        }
+
+                    }
+                } catch (ParseException e) {
+                    command.errors.reject("error.expirationDate.invalid", "Row ${rowIndex}: Product '${row.productCode}' must have a valid date (or no date)")
+                }
+
+
+                def levenshteinDistance = StringUtils.getLevenshteinDistance(product.name, row.product)
+                if (row.product && levenshteinDistance > 0) {
+                    command.warnings[index] << "Product name [${row.product}] does not appear to be the same as in the database [${product.name}] (Levenshtein distance: ${levenshteinDistance})"
+                }
+
+            }
+
+            if (row.quantity && (row.quantity as int) < 0) {
+                command.errors.reject("error.quantity.negative", "Row ${rowIndex}: Product '${row.productCode}' must have positive quantity")
+            }
+        }
+    }
```

Comment:
the complicated logic here won with me :see_no_evil: 
I'll skip that for later
what I can say for now is that at least we should add types where it's possible, instead of spamming def.

Reply:
This is the first place where I think a command object with validation rules makes sense.

Reply:
Also this crap might be a good candidate for cell reporters. 
```
command.warnings[index]
```